reply.lua

```diff
@@ -0,0 +1,166 @@
+--
+-- SPDX-FileCopyrightText: (c) 2024 Ring Zero Desenvolvimento de Software LTDA
+-- SPDX-License-Identifier: GPL-2.0-only
+--
+
+local data   = require("data")
+local linux  = require("linux")
+local socket = require("socket")
+local pack   = require("dome/pack")
+local config = require("dome/config")
+
+local PACKET = socket.af.PACKET
+local RAW    = socket.sock.RAW
+
+local ETH_HLEN  = 14
+local ETH_P_ALL = 0x0003
+
+local IPPROTO_TCP = 0x06
+
+local RST = 1 << 2
+local ACK = 1 << 4
+
+local packer   = pack.packer
+local unpacker = pack.unpacker
+
+local reply = {}
+
+local ntoh16 = linux.ntoh16
+local function checksum(data, base, length)
+	local base = base or 0
+	local length = length or #data - base
+	local sum = 0
+
+	for offset = 0, length - 1, 2 do
+		sum = sum + ntoh16(data:getuint16(base + offset))
+	end
+	if length % 2 == 1 then
+		sum = sum + data:getu8(base + length - 1)
+	end
+	sum = sum & 0xffffffff
+	while sum >> 16 ~= 0 do
+		sum = (sum >> 16) + sum & 0xffff
+	end
+
+	return ~sum & 0xffff
+end
+
+local function ethhdr(packet)
+	local getstr, _, getbe16 = unpacker(packet)
+	return {
+		dst     = getstr(0, 6),
+		src     = getstr(6, 6),
+		ethtype = getbe16(12),
+	}
+end
+
+local function iphdr(packet)
+	local getstr, getu8, getbe16, getbe32 = unpacker(packet, ETH_HLEN)
+	return {
+		version  = getu8(0) >> 4,
+		ihl      = getu8(0) & 0xf,
+		tos      = getu8(1),
+		tot_len  = getbe16(2),
+		id       = getbe16(4),
+		frag_off = getbe16(6),
+		ttl      = getu8(8),
+		protocol = getu8(9),
+		check    = getbe16(10),
+		saddr    = getbe32(12),
+		daddr    = getbe32(16),
+	}
+end
+
+local function tcphdr(packet)
+	local ihl = packet:getbyte(ETH_HLEN) & 0xf
+	local base = ETH_HLEN + ihl * 4
```

Comment:
What's `base` here? Is it the payload offset? If so, I would call it `payload`

Reply:
It's not tough. It's the first byte of the TCP Header.

Reply:
Please notice that `base` is used in:

- `pack`, the original
- `checksum`, meaning the first by to account for the sum

- `tcphdr`, meaning the first byte of the tcphdr
- `tcp_checksum`, same meaning as the above
- `edit_tcp`, same meaning as the above

we could change the last three perhaps, not sure? 
what do you suggest?

Reply:
In those cases, I think offset is clearer.. in pack, I use base because we have both base and offset, in this case it's like the "base offset" and and the actual "call offset".. perhaps it needs to be clarified there as well.. not sure..

Reply:
but it's kind of the same thing here. since `local base = ETH_HLEN + ihl * 4` is the beginning of the TCP Header, all fields are inferenced from that "base". 

what should we do? `s/base/offset`? I'm ok with any option here.

Reply:
IMO, `offset` is better.. but I'm ok with `base` as well.. you can decide =)